v1/v1util/zip.go

```diff
@@ -25,16 +25,17 @@ var gzipMagicHeader = []byte{'\x1f', '\x8b'}
 // GzipReadCloser reads uncompressed input data from the io.ReadCloser and
 // returns an io.ReadCloser from which compressed data may be read.
 func GzipReadCloser(r io.ReadCloser) (io.ReadCloser, error) {
-	defer r.Close()
+	pr, pw := io.Pipe()
 
-	// TODO(mattmoor): How to avoid buffering this whole thing into memory?
-	data := bytes.NewBuffer(nil)
-	gw := gzip.NewWriter(data)
-	defer gw.Close()
-	if _, err := io.Copy(gw, r); err != nil {
-		return nil, err
-	}
-	return NopReadCloser(data), nil
+	go func() {
+		defer pw.Close()
+		defer r.Close()
+		gw, _ := gzip.NewWriterLevel(pw, gzip.BestCompression)
+		defer gw.Close()
+		io.Copy(gw, r)
+	}()
```

Comment:
@ImJasonH Do you have anything better here?

Reply:
If `gzip.NewWriterLevel` returns an error, you can handle it with `pr.CloseWithError(err)`

This means you shouldn't `defer r.Close` until after you've set up the `gzip.Writer`.

If `io.Copy` returns an error, you should at least log it, and possibly pass it...somewhere.

Reply:
> If gzip.NewWriterLevel returns an error, you can handle it with pr.CloseWithError(err)

`gzip.NewWriterLevel` won't return an error when the input is `gzip.BestCompression`
See: https://golang.org/src/compress/gzip/gzip.go?s=1841:1901#L50

> This means you shouldn't defer r.Close until after you've set up the gzip.Writer.

When we close the `io.Reader(r)` shouldn't matter. I'd say what does is closing the `gzip.Writer(gw)` prior to closing the `io.PipeWriter(pw)` so that it flushes.

> If io.Copy returns an error, you should at least log it, and possibly pass it...somewhere.

We'll change this so that if `io.Copy`fails we'll forward that to `pr.CloseWithError(err)`. Thus if copying fails then the pipe is closed.

Does that sound good to everyone?

Reply:
You're right about `NewWriterLevel`, I just hate to see an underscored err :disappointed: -- it's fine in this case though.

What I meant was that, _if_ `NewWriterLevel` could return an error, we should pass that to `pr.CloseWithError` and return from the goroutine, in which case we wouldn't want the `defer` to also be queued up. It doesn't matter though since we're not going to handle the (non-existent) error.

Passing `io.Copy`'s error to `pr.CloseWithError` SGTM.